provider/oci/instance.go

```diff
@@ -304,3 +307,65 @@ func (o *ociInstance) refresh() error {
 	o.raw = response.Instance
 	return nil
 }
+
+// OpenPorts (InstanceFirewaller) ensures that the input ingress rule is
+// permitted for machine with the input ID.
+func (o *ociInstance) OpenPorts(
+	ctx envcontext.ProviderCallContext, _ string, rules []network.IngressRule,
+) error {
+	client, err := o.getInstanceConfigurator(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	return errors.Trace(client.ChangeIngressRules("", true, rules))
+}
+
+// OpenPorts (InstanceFirewaller) ensures that the input ingress rule is
+// restricted for machine with the input ID.
+func (o *ociInstance) ClosePorts(
+	ctx envcontext.ProviderCallContext, _ string, rules []network.IngressRule,
+) error {
+	client, err := o.getInstanceConfigurator(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	return errors.Trace(client.ChangeIngressRules("", false, rules))
+}
+
+// IngressRules (InstanceFirewaller) returns the ingress rules that have been
+// applied to the input machine ID.
+func (o *ociInstance) IngressRules(
+	ctx envcontext.ProviderCallContext, _ string,
+) ([]network.IngressRule, error) {
+	client, err := o.getInstanceConfigurator(ctx)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	rules, err := client.FindIngressRules()
+	return rules, errors.Trace(err)
+}
+
+func (o *ociInstance) getInstanceConfigurator(
+	ctx envcontext.ProviderCallContext,
+) (common.InstanceConfigurator, error) {
+	addresses, err := o.Addresses(ctx)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if len(addresses) == 0 {
```

Comment:
Since iterating an empty slice is a no-op you could probably omit this and just fall through to the error at the end of the method.

Reply:
Yeah, I thought the different log lines were worth having, but if the provider thinks there are no addresses at all, we'd have a problems before getting here.